oak_functions/loader/src/metrics.rs

```diff
@@ -0,0 +1,307 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::Context;
+use rand::{distributions::Open01, rngs::StdRng, thread_rng, Rng, SeedableRng};
+use serde::Deserialize;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex},
+};
+
+/// Configuration for differentially-private metrics reporting.
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PrivateMetricsConfig {
+    /// The privacy budget. See
+    /// https://desfontain.es/privacy/differential-privacy-in-practice.html for more information
+    /// on epsilon-differential privacy and Laplacian noise.
+    pub epsilon: f64,
+    /// The number of requests that will be aggregated into each batch.
+    pub batch_size: usize,
+    /// The labels for events that can be reported. Any other events will be ignored.
+    pub allowed_labels: Vec<String>,
+}
+
+/// Aggregator for count-based differentially private metrics. The request metrics are released in
+/// aggregated batches after differentially private noise has been added. Once the number of
+/// requests reaches the batch threshold the metrics are logged and the request count and bucket
+/// counts are reset to 0.
+pub struct PrivateMetricsAggregator {
+    /// The request count.
+    count: usize,
+    /// The privacy budget.
+    epsilon: f64,
+    /// The number of requests that will be aggregated into each batch.
+    batch_size: usize,
+    /// The storage for event count buckets.
+    events: HashMap<String, usize>,
+    /// The random number generator used for sampling the noise.
+    rng: StdRng,
+}
+
+impl PrivateMetricsAggregator {
+    pub fn new(config: &PrivateMetricsConfig) -> anyhow::Result<Self> {
+        anyhow::ensure!(config.epsilon > 0.0, "Epsilon must be positive.");
+        Ok(Self {
+            count: 0,
+            epsilon: config.epsilon,
+            batch_size: config.batch_size,
+            events: config
+                .allowed_labels
+                .iter()
+                .map(|label| (label.clone(), 0))
+                .collect(),
+            rng: StdRng::from_rng(thread_rng()).context("Couldn't create rng")?,
+        })
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test(config: &PrivateMetricsConfig, rng: StdRng) -> anyhow::Result<Self> {
```

Comment:
It would be nice to share this fn with the regular constructor, since there is some logic going on here; maybe rename to `new_with_rng` or similar, leave it private, then use it from `new` and also from `new_for_test` (or just `cfg(test)` and use `new_with_rng` directly from tests). Up to you.

Reply:
Good idea. Done.